### ISSUE TITLE
React 16 gives unique key warning

### ISSUE THREAD

**Issue by crowdwave** [AllenFang/react-bootstrap-table#1681]
As shown in screenshot.  

Any ideas how to fix this?

![screen shot 2017-10-12 at 12 13 22 pm](https://user-images.githubusercontent.com/13228330/31474517-03321292-af47-11e7-8a4c-9e23683f226e.png)

**Comment by sr1994lu** [comment]
same issue

**Comment by AllenFang** [comment]
HI guy, any example for this issue, I cant reproduce this problems, thanks

**Comment by sr1994lu** [comment]
```
/* @flow */
import * as React from 'react';
import { NavLink } from 'react-router-dom';
import { Map, is } from 'immutable';
import classNames from 'classnames';
import { Container, Row, Col } from 'reactstrap';
import { Input, Loader } from 'semantic-ui-react';
import { BootstrapTable, TableHeaderColumn } from 'react-bootstrap-table';
import { EditPencil } from '../Icons/Icons';

export type Props = {
  isParentDatas: boolean,
  searchPlaceHolder: string,
  masterName: string,
  datasName: string,
  datasIndex: number,
  registrationContent: string,
  datas: [],
  options: {
    page: number,
    defaultSortName: string,
    defaultSortOrder: string,
    prePage: string,
    nextPage: string,
    noDataText: any,
    paginationPosition: string,
    hideSizePerPage: boolean,
    sortIndicator: boolean,
  },
  showEditLink: boolean,
  isLinkColumn: boolean,
  datasHeaderFieldColumns: [],
  showModified: boolean,
  showCreated: boolean,
  showEditLink: boolean,
  datasHeaderWidths: [],
  dispatchEditLink: void,
  dispatchDetailLink: void,
  datasHeaderFieldChildren: string
}

class Table extends React.Component<Props> {
  constructor(props: Props) {
    super(props);
    this.editLink = this.editLink.bind(this);
    this.detailLink = this.detailLink.bind(this);
    this.subscribeColumns = this.subscribeColumns.bind(this);
  }

  shouldComponentUpdate(nextProps, nextState) {
    const nextPropsMap = Map(nextProps);
    const nextStateMap = Map(nextState);
    if (!is(nextPropsMap, this.prevPropsMap) || !is(nextStateMap, this.prevStateMap)) {
      this.prevPropsMap = nextPropsMap;
      this.prevStateMap = nextStateMap;
      return true;
    }
    return false;
  }


  editLink(cell, row, formatExtraData, rowIdx) {
    return (
      <NavLink
        to={`/${this.props.masterName}/edit`}
        onClick={() => this.props.dispatchEditLink}
        key={rowIdx}
      >
        <EditPencil />
      </NavLink>
    );
  }

  detailLink(cell, row, rowIdx) {
    return (
      <NavLink
        to={`/${this.props.masterName}/detail`}
        onClick={() => this.props.dispatchDetailLink}
        key={rowIdx}
      >
        {cell}
      </NavLink>
    );
  }

  subscribeColumns(column, index) {
    return (
      <TableHeaderColumn
        dataField={column}
        dataSort
        width={this.props.datasHeaderWidths[index]}
        dataFormat={(cell, row) => (
            this.props.isLinkColumn === column
              ? this.detailLink(cell, row, index)
              : cell
          )}
        key={index}
      >
        {this.props.datasHeaderFieldChildren[index]}
      </TableHeaderColumn>
    );
  }

  render() {
    const {
      isParentDatas,
      searchPlaceHolder,
      masterName,
      datasName,
      datasIndex,
      registrationContent,
      datas,
      options = {
        page: 1,
        defaultSortName: 'id',
        defaultSortOrder: 'desc',
        prePage: 'before',
        nextPage: 'next',
        noDataText: <Loader active size="tiny" inline />,
        paginationPosition: 'top',
        hideSizePerPage: false,
        sortIndicator: false
      },
      showEditLink,
      datasHeaderFieldColumns,
      showModified,
      showCreated
    } = this.props;
    return (
      <Container className={classNames('common', 'table')}>
        {isParentDatas
        ?
          <Container>
            <Row>
              <Col md={12}>
                <Input
                  label={{ icon: 'search' }}
                  placeholder={searchPlaceHolder}
                />
              </Col>
            </Row>
            <Row>
              <Col md={12}>
                <div>{datasIndex} of views</div>
              </Col>
            </Row>
            <Row>
              <Col md={12}>
                <NavLink to={`/${datasName}/create`}>
                  {`${registrationContent} added`}
                </NavLink>
              </Col>
            </Row>
          </Container>
        :
          <Container>
            <Row>
              <Col md={12}>
                <Input
                  label={{ icon: 'search' }}
                  placeholder={searchPlaceHolder}
                />
              </Col>
            </Row>
            <Row>
              <Col md={12}>
                <div>{datasIndex}件の子ユーザを表示中</div>
              </Col>
            </Row>
          </Container>
        }
        <BootstrapTable
          data={datas}
          pagination
          options={options}
          hover
        >
          {
            showEditLink
              ?
                <TableHeaderColumn
                  dataField=""
                  width="30"
                  dataAlign="center"
                  dataFormat={this.editLink}
                />
              :
                <TableHeaderColumn
                  dataField=""
                  width="1"
                  dataAlign="center"
                  dataFormat={this.editLink}
                />
          }
          {
            datasHeaderFieldColumns.map((elm, idx) => this.subscribeColumns(elm, idx))
          }
          {
            showModified
              ?
                <TableHeaderColumn
                  dataField="modifiedAt"
                  dataSort
                  width="120"
                >
                  更新日
                </TableHeaderColumn>
              :
                <TableHeaderColumn
                  dataField="modifiedAt"
                  dataSort
                  width="1"
                >
                  {' '}
                </TableHeaderColumn>
          }
          {
            showCreated
              ?
                <TableHeaderColumn
                  dataField="createdAt"
                  dataSort
                  width="120"
                >
                  {
                    masterName === 'users'
                      ? 'Contract date'
                      : 'Created date'
                  }
                </TableHeaderColumn>
              :
                <TableHeaderColumn
                  dataField="createdAt"
                  dataSort
                  width="1"
                >
                  {' '}
                </TableHeaderColumn>
          }
          <TableHeaderColumn
            dataField="id"
            dataSort
            width="1"
            isKey
          >
            {' '}
          </TableHeaderColumn>
        </BootstrapTable>
      </Container>
    );
  }
}

export default Table;

```

**Comment by AllenFang** [comment]
https://reactjs.org/blog/2017/09/26/react-v16.0.html#new-render-return-types-fragments-and-strings

**Comment by sr1994lu** [comment]
@AllenFang 
this link is not such error.

I have:
Warning: Each child in an array or iterator should have a unique "key" prop. See https://fb.me/react-warning-keys for more information.

<img width="869" alt="2017-10-16_17-51-16" src="https://user-images.githubusercontent.com/17734462/31603298-a9b8b89a-b29a-11e7-9621-7e3ef71fe9f3.png">

**Comment by crowdwave** [comment]
Hi Allen

I'm not sure I understand... can you suggest what I need to do to fix this?

thanks

**Comment by AllenFang** [comment]
@crowdwave could you have a most simplest example for reproducing this issue?

**Comment by AllenFang** [comment]
BTW, you guy can upgrade to `v4.1.0` and try again.

**Comment by crowdwave** [comment]
I found the problem. In my case, it was because I had prerendered JSX that I was storing and just returning that from the column formatter function.  This worked in React15 and for some reason it is not valid in React16, so I have removed my code that does that.

 @sr1994lu I suggest to fix your problem, you remove all columns from the table then add them back one by one until you find each column causing a problem then examine that column in depth to find out what it is about the data or the function that is a problem.  Happy to help further if you do that.

thanks
